### ISSUE TITLE
3x speedup compression claim is false/outdated

### ISSUE THREAD

**Issue by TyrelCB** [lyogavin/airllm#330]
Final: the compression claim is inverted on your GB10

Qwen2.5-32B-Instruct, 10 tokens, same prompt, run sequentially:

┌──────┬─────────┬─────────────┬──────────┬─────────────┬───────┐
│ mode │ s/token │ vs baseline │ peak GPU │ bytes/layer │ split │
├──────┼─────────┼─────────────┼──────────┼─────────────┼───────┤
│ none │ 13.3    │ —           │ 1.60GB   │ 975MB       │ 62G   │
├──────┼─────────┼─────────────┼──────────┼─────────────┼───────┤
│ 8bit │ 50.5    │ 3.8x slower │ 3.16GB   │ 489MB       │ 31G   │
├──────┼─────────┼─────────────┼──────────┼─────────────┼───────┤
│ 4bit │ 106.7   │ 8.0x slower │ 2.04GB   │ 274MB       │ 18G   │
└──────┴─────────┴─────────────┴──────────┴─────────────┴───────┘

All three produced correct output — speed and memory only, no accuracy issue.

Against the repo: README:142 claims "up to 3x" speedup; utils.py:498 claims "4x or 2x". Measured: 8x and 3.8x slower, and every compressed mode used more memory than baseline. Plus a one-time 361s split build for 4bit.

Why, and the 4bit run is what proves it: 4bit reads the least data — 274MB/layer, 3.6x less than baseline, entire 18GB split fitting in your 76GB page cache. If the win were I/O, it should have been fastest. It was slowest. Speed tracks dequantization work, not bytes read. uncompress_layer_state_dict (utils.py:86-114) expands every layer back to fp16 per token — files shrink, compute doesn't. 4bit's NF4 path additionally rebuilds a QuantState per tensor through the O(n²) key scan at line 92, costing more than 8bit's simpler path despite reading less.

I'd call this a stale 2023 benchmark, not a dishonest one. On airllm 2.0-era disks, paying dequant cost to read a third of the bytes was plausibly a real win. On NVMe witre nearly free and that overhead is pure loss.

These numbers cleanly contradict a headline README claim, with the causal code path identified. Happy to write it up.

**Comment by seaurching** [comment]
这个运行起来要花费多久才能输出。我修改了一下适配kimi2.5它要花费25分钟才能得到结果
```
from airllm import AutoModel

MAX_LENGTH = 128
model = AutoModel.from_pretrained("/DATA/disk1/moonshotai/Kimi-K2.5")

input_text = [
        'What is the capital of United States?',

[airllvm_fix_report.md](https://github.com/user-attachments/files/30731797/airllvm_fix_report.md)

    ]

input_tokens = model.tokenizer(input_text,
    return_tensors="pt", 
    return_attention_mask=False, 
    truncation=True, 
    max_length=MAX_LENGTH, 
    padding=False)
           
generation_output = model.generate(
    input_tokens['input_ids'].cuda(),
    max_new_tokens=20,
    use_cache=True,
    return_dict_in_generate=True)

output = model.tokenizer.decode(generation_output.sequences[0].tolist())

print(output)
```
<img width="434" height="192" alt="Image" src="https://github.com/user-attachments/assets/c51f7bd3-b123-4df8-934d-0c645db335c3" />

**Comment by rodriguezrussell93** [comment]
📌 消息队列至少一次还是恰好一次

**Comment by Hotragn** [comment]
Your numbers aren't a GB10 quirk — the direction is **structural**, and it's visible in the source. Two things happen when you pass `compression=`, and both cost time on the hot path:

**1. Compression force-disables prefetching** — `airllm_base.py:172-174`:

```python
if self.compression is not None and self.prefetching:
    print("prefetching is not supported together with compression for now; disabling prefetching.")
    self.prefetching = False
```

Prefetching is what overlaps the next layer's load with the current layer's compute (the "10% speed improvement" from the v2.5 notes). With compression on you lose that overlap entirely, so every layer load serializes against compute.

**2. Every layer gets dequantized on the GPU, per layer, per token.** `load_layer` → `uncompress_layer_state_dict` (`utils.py:142`, `utils.py:86-114`). For 4-bit, per tensor per layer:

```python
quant_state = bnb.functional.QuantState.from_dict(qs_dict=quant_state_dict, device="cuda")
dqv = bnb.functional.dequantize_nf4(v.cuda(), quant_state)
```

So you pay a `.cuda()` upload of the quantized tensor plus a dequant kernel plus `QuantState` reconstruction — for all ~64 layers on every token. 8-bit takes the cheaper `dequantize_blockwise` path, which is why **8bit lands between none and 4bit in your table** rather than the two compressed modes being similar.

That also explains the peak-GPU inversion you measured (8bit 3.16GB vs none 1.60GB), which is otherwise counterintuitive: dequantize materializes the fp16 tensor on the GPU *alongside* the quantized source and its quant state, so the compressed path transiently holds more than the uncompressed one.

**Why the claim was ever true:** the trade is disk bytes for GPU work — your 62G→18G split shrink is real. That's a win only when disk read is the bottleneck, which it was on the 2023-era commodity setup the "3x" came from. On a GB10 with fast unified memory, disk was never the constraint, so it's pure overhead. Both can be true; the README just states it unconditionally.

Caveat on my end: I verified the mechanism by reading the code, not by reproducing your benchmark — no GB10 here, and my box is CPU-only. So treat the mechanism as confirmed and the magnitudes as yours.

This is also the second time the headline claims have come up without reproducible evidence attached (#295, and I proposed a benchmark harness in #312 partly for this). Your table is the most concrete data anyone's put on it — it'd be a shame for it to sit unaddressed. If it's useful, I'm happy to help turn it into a pinned-revision benchmark script under `benchmarks/` so the claim becomes falsifiable per-hardware instead of absolute.

**Comment by matthewsdavid456078** [comment]
破坏性变更提前多久公示 🤝🤝
